test-e2e/test_zookeeper_backup.py

```diff
@@ -0,0 +1,283 @@
+    """
+    zk_hostports = ','.join(['{}:2181'.format(m.public_ip_address)
+                             for m in three_master_cluster.masters])
+    zk_client = KazooClient(
+        hosts=zk_hostports,
+        # Avoid failure due to client session timeout.
+        timeout=60,
+    )
+    def zk_listener(state: KazooState) -> None:
+        # Open session after temporary connection loss.
+        if state == KazooState.CONNECTED:
+            zk_client.start()
+
+    zk_client.add_listener(zk_listener)
+    zk_client.start()
+    try:
+        yield zk_client
+    finally:
+        zk_client.stop()
+        zk_client.close()
+
+
+def _zk_set_flag(zk_client: KazooClient, ephemeral: bool = False) -> str:
+    """
+    Store the `FLAG` value in ZooKeeper in a random Znode.
+    """
+    znode = '/{}'.format(uuid.uuid4())
+    zk_client.create(znode, makepath=True, ephemeral=ephemeral)
+    zk_client.set(znode, FLAG)
+    return znode
+
+
+def _zk_flag_exists(zk_client: KazooClient, znode: str) -> bool:
+    """
+    The `FLAG` value exists in ZooKeeper at `znode` path.
+    """
+    try:
+        value = zk_client.get(znode)
+    except NoNodeError:
+        return False
+    return bool(value[0] == FLAG)
+
+
+class TestZooKeeperBackup:
+    """
+    Within the context of DC/OS ZooKeeper can be backed up on a running
+    cluster and a previous state can be restored.
+    """
+
+    def test_transaction_log_backup_and_restore(
+        self,
+        three_master_cluster: Cluster,
+        zk_client: KazooClient,
+        tmp_path: Path,
+        request: SubRequest,
+        log_dir: Path,
+    ) -> None:
+        """
+        In a 3-master cluster, backing up the transaction log of ZooKeeper on
+        one node and restoring from the backup on all master results in a
+        functioning DC/OS cluster with previously backed up Znodes restored.
+        """
+        # Write to ZooKeeper before backup
+        persistent_flag = _zk_set_flag(zk_client)
+        ephemeral_flag = _zk_set_flag(zk_client, ephemeral=True)
+
+        random = uuid.uuid4().hex
+        backup_name = 'zk-backup-{random}.tar.gz'.format(random=random)
+        backup_local_path = tmp_path / backup_name
+
+        # Take ZooKeeper backup from one master node.
+        _do_backup(next(iter(three_master_cluster.masters)), backup_local_path)
+
+        # Store a datapoint which we expect to get lost.
+        not_backed_up_flag = _zk_set_flag(zk_client)
+
+        # Restore ZooKeeper from backup on all master nodes.
+        _do_restore(three_master_cluster.masters, backup_local_path)
+
+        # Assert that DC/OS is intact.
+        wait_for_dcos_oss(
+            cluster=three_master_cluster,
+            request=request,
+            log_dir=log_dir,
+        )
+
+        # Read from ZooKeeper after restore
+        assert _zk_flag_exists(zk_client, persistent_flag)
+        assert _zk_flag_exists(zk_client, ephemeral_flag)
```

Comment:
How do we know that the restore and `wait_for_dcos_oss` took less than the ephemeral flag timeout?

Reply:
What do you mean by the `ephemeral flag timeout`?
In my understanding there are two timeouts active here:
* Client connection timeout (60 seconds, set above). This needs to be longer than the restore procedure (we can set it arbitrarily)
* ZooKeeper session timeout -> resets after ZooKeeper comes back alive. Default is 10 seconds.
The client would need to reconnect within 10 seconds after ZooKeeper comes up again.

Reply:
As it appears this was the culprit! Thanks Jon for raising this in the first place :100: